actors/cache/impl/zcache.go

```diff
@@ -0,0 +1,211 @@
+	shortCidMap    zcache.ZCache
+	robustShortMap zcache.ZCache
+	shortRobustMap zcache.ZCache
+	logger         *zap.Logger
+	cacheType      string
+}
+
+func (m *ZCache) NewImpl(source common.DataSource, logger *zap.Logger) error {
+	var err error
+	m.logger = logger2.GetSafeLogger(logger)
+
+	// If no config was provided, the combined cache is configured as
+	// remote best effort, as the remote cache will fail. However, the cache will
+	// work anyway
+	cacheConfig := source.Config.Cache
+	if cacheConfig == nil {
+		m.cacheType = "in-memory"
+		if m.shortCidMap, err = zcache.NewLocalCache(&zcache.LocalConfig{Prefix: Short2CidMapPrefix}); err != nil {
+			return err
+		}
+		if m.robustShortMap, err = zcache.NewLocalCache(&zcache.LocalConfig{Prefix: Robust2ShortMapPrefix}); err != nil {
+			return err
+		}
+		if m.shortRobustMap, err = zcache.NewLocalCache(&zcache.LocalConfig{Prefix: Short2RobustMapPrefix}); err != nil {
+			return err
+		}
+	} else {
+		m.cacheType = "combined"
+
+		prefix := ""
+		if cacheConfig.GlobalPrefix != "" {
+			prefix = fmt.Sprintf("%s/", cacheConfig.GlobalPrefix)
+		}
+		if source.Config.NetworkName != "" {
+			prefix = fmt.Sprintf("%s%s/", prefix, source.Config.NetworkName)
+		}
+
+		shortCidMapConfig := &zcache.CombinedConfig{
+			GlobalPrefix:       fmt.Sprintf("%s%s", prefix, Short2CidMapPrefix),
+			GlobalTtlSeconds:   cacheConfig.GlobalTtlSeconds,
+			IsRemoteBestEffort: cacheConfig.IsRemoteBestEffort,
+			Local:              cacheConfig.Local,
+			Remote:             cacheConfig.Remote,
+		}
+
+		robustShortMapConfig := &zcache.CombinedConfig{
+			GlobalPrefix:       fmt.Sprintf("%s%s", prefix, Robust2ShortMapPrefix),
+			GlobalTtlSeconds:   -1,
+			IsRemoteBestEffort: cacheConfig.IsRemoteBestEffort,
+			Local:              cacheConfig.Local,
+			Remote:             cacheConfig.Remote,
+		}
+
+		shortRobustMapConfig := &zcache.CombinedConfig{
+			GlobalPrefix:       fmt.Sprintf("%s%s", prefix, Short2RobustMapPrefix),
+			GlobalTtlSeconds:   -1,
+			IsRemoteBestEffort: cacheConfig.IsRemoteBestEffort,
+			Local:              cacheConfig.Local,
+			Remote:             cacheConfig.Remote,
+		}
+
+		if m.shortCidMap, err = zcache.NewCombinedCache(shortCidMapConfig); err != nil {
+			return err
+		}
+		if m.robustShortMap, err = zcache.NewCombinedCache(robustShortMapConfig); err != nil {
+			return err
+		}
+		if m.shortRobustMap, err = zcache.NewCombinedCache(shortRobustMapConfig); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *ZCache) ImplementationType() string {
+	return ZCacheImpl + "/" + m.cacheType
+}
+
+func (m *ZCache) BackFill() error {
+	// Nothing to do
+	return nil
+}
+
+func (m *ZCache) GetActorCode(address address.Address, key filTypes.TipSetKey) (string, error) {
+	shortAddress, err := m.GetShortAddress(address)
+	if err != nil {
+		m.logger.Sugar().Infof("[ActorsCache] - Error getting short address: %s\n", err.Error())
+		return cid.Undef.String(), common.ErrKeyNotFound
```

Comment:
Adding context about which key was not found could be helpful for debugging

Reply:
I like it!